llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -2491,17 +2491,19 @@ void InnerLoopVectorizer::emitIterationCountCheck(BasicBlock *Bypass) {
     Value *LHS = Builder.CreateSub(MaxUIntTripCount, Count);
 
     Value *Step = CreateStep();
-#ifndef NDEBUG
     ScalarEvolution &SE = *PSE.getSE();
     const SCEV *TC2OverflowSCEV = SE.applyLoopGuards(SE.getSCEV(LHS), OrigLoop);
-    assert(
-        !isIndvarOverflowCheckKnownFalse(Cost, VF * UF) &&
-        !SE.isKnownPredicate(CmpInst::getInversePredicate(ICmpInst::ICMP_ULT),
-                             TC2OverflowSCEV, SE.getSCEV(Step)) &&
-        "unexpectedly proved overflow check to be known");
-#endif
-    // Don't execute the vector loop if (UMax - n) < (VF * UF).
-    CheckMinIters = Builder.CreateICmp(ICmpInst::ICMP_ULT, LHS, Step);
+    const SCEV *StepSCEV = SE.getSCEV(Step);
+
+    // Check if (UMax - n) < (VF * UF).
+    if (SE.isKnownPredicate(ICmpInst::ICMP_ULT, TC2OverflowSCEV, StepSCEV)) {
```

Comment:
The patch to add the NDEBUG code landed only recently - https://github.com/llvm/llvm-project/pull/111310 by @fhahn. It looks like this assert was added for a reason - is it possible that `isIndvarOverflowCheckKnownFalse` is returning the incorrect answer for the EVL case?

Reply:
Ah this is interesting and thanks for explaining and putting up a fix! So the test you've added shows `isIndvarOverflowCheckKnownFalse` is returning the wrong answer, or at least an answer that's inconsistent with the `SE.isKnownPredicate` call. Once we've applied the loop guards we know that the check is false. So it looks like there are a few choices:

1. In the assert above we could avoid applying the loop guards and then `isIndvarOverflowCheckKnownFalse` will be consistent with the call to `SE.isKnownPredicate`. That might be a much smaller change than this patch I think? Or,
2. Teach `isIndvarOverflowCheckKnownFalse` to apply loop guards, which I think would require changing ScalarEvolution::getSmallConstantMaxTripCount to apply loop guards. It looks like `ScalarEvolution::getSmallConstantTripMultiple` already does something similar. Or,
3. Use the current version of your patch.

The only problem with 3 is that the code is now more complex and it doesn't look like there are any tests for the `CheckMinIters = Builder.getTrue();` case you've added.

I personally feel like 2 is the most powerful change because you've highlighted a problem in how we estimate the small constant max trip count. If the vectoriser is always going to apply loop guards anyway, then we should do the same when estimating the trip count. However, I realise it would expand the scope of this patch quite a bit and I perfectly understand if you don't have time to do this right now! What do you think?

Reply:
This is an interesting case. For the loop in the test case, the induction can  overflow if `%tc == 0` I think, so `isIndvarOverflowCheckKnownFalse` should return false in general?

I think `getSmallConstantMaxTripCount` already should make use of loop guards in this case, but we cannot compute a tighter maximum because `%tc` may be 0, in which case we would wrap.

Reply:
Thanks for the review and pointers! 2 seems good to me too, I'm happy to give it a stab.

> This is an interesting case. For the loop in the test case, the induction can overflow if `%tc == 0` I think, so `isIndvarOverflowCheckKnownFalse` should return false in general?

Oh woops, I think the exit condition in the test case should have been `icmp uge`. 

But if we did emit the overflow check would it even have caught it at runtime? If `%tc == 0` then I think `(UMax - %tc) < (VF * UF)` will always return false.

Reply:
I took a look at teaching `isIndvarOverflowCheckKnownFalse` but as @fhahn said I think it's already doing the correct thing. `getConstantMaxBackedgeTakenCount` was returning 0xFFFFFFFFFFFFFFFF due to the overflow case at `%tc == 0`. But `getConstantTripCount` throws it away because it only considers trip counts that fit into 32 bits. 

The SCEV assertion uses the full 64 bits of the TC type though, and in combination with what I think is a separate coincidental issue that `(UMax - %tc) < (VF * UF)` when `%tc == 0` means, it means that the predicate is always true.

I couldn't think of an easy way of updating the assertion, so I've reworked this PR to just remove the assert and nothing else for now.

I've added some new cases to show things that I think could be split off from this PR:

- `@overflow_at_0` shows how the overflow check seems to let `%tc == 0` slip through, which maybe is something that needs fixed in a separate PR?
- `@no_overflow_at_0` shows a case where `isIndvarOverflowCheckKnownFalse` works and knows that the maximum trip count is 1025, so nothing needs to be done here
- `@trip_count_max_1024` shows a case where we currently can't calculate the max trip count, i.e. the debug output doesn't contain:`LV: Found maximum trip count: N`, and so `isIndvarOverflowCheckKnownFalse` also returns false. I think in a separate PR we can teach `computeMaxBECountForLT` to apply the loop guards, which seems to fix this.